Add EasyAnimateV5.1 text-to-video, image-to-video, control-to-video generation model

bubbliiiing opened the pull request:
# What does this PR do?
This PR converts the EasyAnimateV5.1 model into a diffuser-supported inference model, including three complete pipelines and corresponding modules. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@a-r-r-o-w

a-r-r-o-w commented on src/diffusers/models/autoencoders/autoencoder_kl_magvit.py:
```diff
@@ -0,0 +1,1640 @@
+    @register_to_config
+    def __init__(
+        self,
+        in_channels: int = 3,
+        out_channels: int = 3,
+        ch =  128,
+        ch_mult = [ 1,2,4,4 ],
+        block_out_channels = [128, 256, 512, 512],
+        down_block_types: tuple = None,
+        up_block_types: tuple = None,
+        mid_block_type: str = "MidBlock3D",
+        layers_per_block: int = 2,
+        act_fn: str = "silu",
+        latent_channels: int = 4,
+        norm_num_groups: int = 32,
+        scaling_factor: float = 0.1825,
+        force_upcast: float = True,
+        use_tiling=False,
+        mini_batch_encoder=9,
+        mini_batch_decoder=3,
+        spatial_group_norm=False,
+        tile_sample_min_size=384,
+        tile_overlap_factor=0.25,
```
Let's make sure to default these values to the ones used by the VAE. Basically, if we create an instance of this module without passing parameters (like `model = AutoencoderKLMagvit()`), then the state dict should be perfectly compatible with the actual model weights

a-r-r-o-w commented on src/diffusers/models/downsampling.py:
```diff
@@ -353,6 +354,114 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return x
 
 
+class EasyAnimateDownsampler3D(nn.Module):
```
Let's move this to the autoencoder file implementation as we are moving towards having entire model implementation in single file

a-r-r-o-w commented on src/diffusers/models/normalization.py:
```diff
@@ -588,6 +588,121 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         return F.normalize(hidden_states, p=self.p, dim=self.dim, eps=self.eps)
 
 
+class EasyAnimateRMSNorm(nn.Module):
```
Let's move both these normalization layers to the specific file they are used in. If they are used in both VAE and Transformer, then it is okay to keep here

bubbliiiing replied:
Already modified.

a-r-r-o-w reviewed (commented):
Thank you for the PR @bubbliiiing! This is in great shape and already mostly in the implementation style used in diffusers :hugs:

I've left some comments from a quick look through the PR. Happy to help make any of the required changes to help bring the PR to completion

bubbliiiing commented:
Sorry for not standardizing some parts; I will make the necessary modifications. Also, I would like to ask if I need to add test files in `tests/pipelines` and add documentation in `docs/source/en/api/pipelines`?

a-r-r-o-w commented:
Yes, we will need a test for all three pipelines as well as model tests in `tests/models`, as well as the documentation pages. Ofcourse, I will try to help you with everything :)

Also, congratulations on the release! I tried out the original repository example and the model is very good! 🎉

bubbliiiing commented on src/diffusers/models/autoencoders/autoencoder_kl_magvit.py:
```diff
@@ -0,0 +1,1640 @@
+        for name, module in self.named_children():
+            _set_3dgroupnorm_for_submodule(name, module)
+
+    def single_forward(self, x: torch.Tensor) -> torch.Tensor:
```
Sorry, does this mean that I cannot use functions like set_padding_one_frame?

bubbliiiing replied:
Do I need to use this conv_cache in [autoencoder_kl_mochi](https://github.com/huggingface/diffusers/blob/8d6f6d6b664c32db457039ab01b1145a36efd038/src/diffusers/models/autoencoders/autoencoder_kl_mochi.py#L109C9-L109C19)?

a-r-r-o-w replied:
In the model implementations, we usually only try to keep (atleast in the latest model integrations):
- Submodel initializations
- Forward method

So, unless a helper function like `set_padding_one_frame` is used in multiple locations, I would suggest directly substituting its code in the forward implementation. If a helper function is required, let's make it a private function by prefixing the function name with an underscore

The conv_cache saves a few computations when running the VAE encode/decode process from repeated frames that are used as padding. As such, it is not required to implement it if it is not needed for framewise encoding and decoding.

bubbliiiing commented on src/diffusers/models/attention_processor.py:
```diff
@@ -3613,6 +3613,109 @@ def __call__(
+        encoder_hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        image_rotary_emb: Optional[torch.Tensor] = None,
+        attn2: Attention = None,
```
Does this mean using add_q, add_k, add_v instead of using attn2?
![image](https://github.com/user-attachments/assets/26913a6d-6f82-41cd-90e4-5ae87d80286a)

a-r-r-o-w replied:
There are two things we could do here:
- Either convert the state dict of the original-format models (that you currently have on the HuggingFace Hub) and update them to diffusers-format (which would make `attn2.to_q -> add_q_proj`, `attn2.to_k -> add_k_proj`, `attn2.to_v -> add_v_proj`
- Create a custom attention class similar to `Attention` and `MochiAttention` in which you are free to use layer naming of your choice (so basically keeping the same `to_q`, `to_k` and `to_v`.

The first approach is more closely aligned with diffusers code style but would require you to update multiple checkpoints -- but we are transitioning to a single file modeling format, so if you choose to go with second approach for convenience, that works for us as well. Essentially, irrespective of the design you choose, we need to make sure:
- When the forward of a layer is called, it only takes tensors as input and produces tensors as output.
- Taking intermediate layers as input to forward, or making calls to other layers out-of-order randomly, is not supported by our design style of different current/upcoming features

cc @DN6 here in case you have thoughts about the single file format and model-specific Attention classes

bubbliiiing replied:
I moved attn2 to the processor's init; does this meet the requirements?

bubbliiiing commented on .gitignore:
```diff
@@ -1,4 +1,5 @@
 # Initially taken from GitHub's Python gitignore file
+__*/
```
Already modified.

bubbliiiing commented on src/diffusers/models/attention_processor.py:
```diff
@@ -3613,6 +3613,109 @@ def __call__(
         return hidden_states
 
 
+class EasyAnimateAttnProcessor2_0:
```
Already modified.

bubbliiiing commented on src/diffusers/models/autoencoders/autoencoder_kl_magvit.py:
```diff
@@ -0,0 +1,1640 @@
+        stride=1, # : int | tuple[int, int, int] = 1,
+        padding=1, # : int | tuple[int, int, int],  # TODO: change it to 0.
+        dilation=1, # :  int | tuple[int, int, int] = 1,
+        **kwargs,
```
Already modified.

bubbliiiing commented on src/diffusers/models/autoencoders/autoencoder_kl_magvit.py:
```diff
@@ -0,0 +1,1640 @@
+        # Apply group normalization and activation function
+        if self.set_3dgroupnorm:
+            batch_size = x.shape[0]
+            x = rearrange(x, "b c t h w -> (b t) c h w")
```
Already modified.

bubbliiiing commented on src/diffusers/models/autoencoders/autoencoder_kl_magvit.py:
```diff
@@ -0,0 +1,1640 @@
+        return x
+
+
+class SpatialTemporalDownBlock3D(nn.Module):
```
Already modified.

bubbliiiing commented on src/diffusers/models/autoencoders/autoencoder_kl_magvit.py:
```diff
@@ -0,0 +1,1640 @@
+
+        return x
+
+def get_mid_block(
```
Already modified.

a-r-r-o-w commented:
Thank you so much for addressing the reviews @bubbliiiing! The PR is almost ready to merge IMO. There are just a few more small changes to make to align to our model implementation design. Is it okay if I quickly push some last changes to this branch directly?

bubbliiiing commented:
> Thank you so much for addressing the reviews @bubbliiiing! The PR is almost ready to merge IMO. There are just a few more small changes to make to align to our model implementation design. Is it okay if I quickly push some last changes to this branch directly?

Of course, thank you very much for your help.

nitinmukesh commented:
Hello @bubbliiiing 

Is there any possibility to support pipeline.enable_sequential_cpu_offload(). Even with enable_model_cpu_offload, vae slicing/tiling and nf4 quantization it goes out of memory on 16 GB with default settings (512x512)

nitinmukesh commented:
For some reason this isn't working anymore, it was working earlier. :(

```
import torch
from diffusers import BitsAndBytesConfig as DiffusersBitsAndBytesConfig, EasyAnimateTransformer3DModel, EasyAnimatePipeline
from diffusers.utils import export_to_video
from transformers import AutoModel
dtype = torch.bfloat16
quant_config = DiffusersBitsAndBytesConfig(load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=dtype)
text_encoder_4bit = AutoModel.from_pretrained(
    "alibaba-pai/EasyAnimateV5.1-12b-zh",
    subfolder="text_encoder",
    quantization_config=quant_config,
    torch_dtype=torch.bfloat16,
)
transformer_4bit = EasyAnimateTransformer3DModel.from_pretrained(
    "alibaba-pai/EasyAnimateV5.1-12b-zh",
    subfolder="transformer",
    quantization_config=quant_config,
    torch_dtype=dtype,
)
pipeline = EasyAnimatePipeline.from_pretrained(
    "alibaba-pai/EasyAnimateV5.1-12b-zh",
    text_encoder=text_encoder_4bit,
    transformer=transformer_4bit,
    torch_dtype=dtype,
)
pipeline.enable_model_cpu_offload()
pipeline.vae.enable_slicing()
pipeline.vae.enable_tiling()
prompt = "A cat walks on the grass, realistic style."
negative_prompt = "bad detailed"
video = pipeline(
    prompt=prompt, 
    negative_prompt=negative_prompt, 
    num_frames=81, 
    num_inference_steps=40,
    width=512,
    height=320
).frames[0]
export_to_video(video, "cat2.mp4", fps=8)

```

It just crashes

```
(venv) C:\aiOWN\lumina2>python EasyAnimate5.1.py
`low_cpu_mem_usage` was None, now default to True since model is quantized.
Downloading shards: 100%|███████████████████████████████████████████████████| 5/5 [00:00<?, ?it/s]
`Qwen2VLRotaryEmbedding` can now be fully parameterized by passing the model config through the `config` argument. All other arguments will be removed in v4.46
Loading checkpoint shards: 100%|████████████████████████████████████| 5/5 [00:19<00:00,  3.90s/it]
The config attributes {'add_ref_latent_in_control_model': True, 'clip_channels': None, 'enable_clip_in_inpaint': False, 'ref_channels': None, 'swa_layers': None} were passed to EasyAnimateTransformer3DModel, but are not expected and will be ignored. Please verify your config.json configuration file.

```

bubbliiiing commented:
> 由于某种原因，它不再起作用了，之前它是起作用的。:(
> 
> ```
> import torch
> from diffusers import BitsAndBytesConfig as DiffusersBitsAndBytesConfig, EasyAnimateTransformer3DModel, EasyAnimatePipeline
> from diffusers.utils import export_to_video
> from transformers import AutoModel
> dtype = torch.bfloat16
> quant_config = DiffusersBitsAndBytesConfig(load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=dtype)
> text_encoder_4bit = AutoModel.from_pretrained(
>     "alibaba-pai/EasyAnimateV5.1-12b-zh",
>     subfolder="text_encoder",
>     quantization_config=quant_config,
>     torch_dtype=torch.bfloat16,
> )
> transformer_4bit = EasyAnimateTransformer3DModel.from_pretrained(
>     "alibaba-pai/EasyAnimateV5.1-12b-zh",
>     subfolder="transformer",
>     quantization_config=quant_config,
>     torch_dtype=dtype,
> )
> pipeline = EasyAnimatePipeline.from_pretrained(
>     "alibaba-pai/EasyAnimateV5.1-12b-zh",
>     text_encoder=text_encoder_4bit,
>     transformer=transformer_4bit,
>     torch_dtype=dtype,
> )
> pipeline.enable_model_cpu_offload()
> pipeline.vae.enable_slicing()
> pipeline.vae.enable_tiling()
> prompt = "A cat walks on the grass, realistic style."
> negative_prompt = "bad detailed"
> video = pipeline(
>     prompt=prompt, 
>     negative_prompt=negative_prompt, 
>     num_frames=81, 
>     num_inference_steps=40,
>     width=512,
>     height=320
> ).frames[0]
> export_to_video(video, "cat2.mp4", fps=8)
> ```
> 
> 它只是崩溃了
> 
> ```
> (venv) C:\aiOWN\lumina2>python EasyAnimate5.1.py
> `low_cpu_mem_usage` was None, now default to True since model is quantized.
> Downloading shards: 100%|███████████████████████████████████████████████████| 5/5 [00:00<?, ?it/s]
> `Qwen2VLRotaryEmbedding` can now be fully parameterized by passing the model config through the `config` argument. All other arguments will be removed in v4.46
> Loading checkpoint shards: 100%|████████████████████████████████████| 5/5 [00:19<00:00,  3.90s/it]
> The config attributes {'add_ref_latent_in_control_model': True, 'clip_channels': None, 'enable_clip_in_inpaint': False, 'ref_channels': None, 'swa_layers': None} were passed to EasyAnimateTransformer3DModel, but are not expected and will be ignored. Please verify your config.json configuration file.
> ```

Thank you for your feedback, I'll give it a try.

bubbliiiing commented:
> For some reason this isn't working anymore, it was working earlier. :(
> 
> ```
> import torch
> from diffusers import BitsAndBytesConfig as DiffusersBitsAndBytesConfig, EasyAnimateTransformer3DModel, EasyAnimatePipeline
> from diffusers.utils import export_to_video
> from transformers import AutoModel
> dtype = torch.bfloat16
> quant_config = DiffusersBitsAndBytesConfig(load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=dtype)
> text_encoder_4bit = AutoModel.from_pretrained(
>     "alibaba-pai/EasyAnimateV5.1-12b-zh",
>     subfolder="text_encoder",
>     quantization_config=quant_config,
>     torch_dtype=torch.bfloat16,
> )
> transformer_4bit = EasyAnimateTransformer3DModel.from_pretrained(
>     "alibaba-pai/EasyAnimateV5.1-12b-zh",
>     subfolder="transformer",
>     quantization_config=quant_config,
>     torch_dtype=dtype,
> )
> pipeline = EasyAnimatePipeline.from_pretrained(
>     "alibaba-pai/EasyAnimateV5.1-12b-zh",
>     text_encoder=text_encoder_4bit,
>     transformer=transformer_4bit,
>     torch_dtype=dtype,
> )
> pipeline.enable_model_cpu_offload()
> pipeline.vae.enable_slicing()
> pipeline.vae.enable_tiling()
> prompt = "A cat walks on the grass, realistic style."
> negative_prompt = "bad detailed"
> video = pipeline(
>     prompt=prompt, 
>     negative_prompt=negative_prompt, 
>     num_frames=81, 
>     num_inference_steps=40,
>     width=512,
>     height=320
> ).frames[0]
> export_to_video(video, "cat2.mp4", fps=8)
> ```
> 
> It just crashes
> 
> ```
> (venv) C:\aiOWN\lumina2>python EasyAnimate5.1.py
> `low_cpu_mem_usage` was None, now default to True since model is quantized.
> Downloading shards: 100%|███████████████████████████████████████████████████| 5/5 [00:00<?, ?it/s]
> `Qwen2VLRotaryEmbedding` can now be fully parameterized by passing the model config through the `config` argument. All other arguments will be removed in v4.46
> Loading checkpoint shards: 100%|████████████████████████████████████| 5/5 [00:19<00:00,  3.90s/it]
> The config attributes {'add_ref_latent_in_control_model': True, 'clip_channels': None, 'enable_clip_in_inpaint': False, 'ref_channels': None, 'swa_layers': None} were passed to EasyAnimateTransformer3DModel, but are not expected and will be ignored. Please verify your config.json configuration file.
> ```

It seems to be working fine in my environment.
![image](https://github.com/user-attachments/assets/79442f1f-cfe7-4355-8d50-9fee3740e712)

nitinmukesh commented:
@bubbliiiing Sorry for the trouble.
It's working now after reinstalling the PR.

nitinmukesh commented:
https://github.com/user-attachments/assets/55c8b792-944f-4574-8aeb-a5f63b094c8e


A little white rabbit with glasses was sitting on a chair in a cafe reading a newspaper. There was a cup of hot coffee on the table.

Nice output. Please add support of pipeline.enable_sequential_cpu_offload().

a-r-r-o-w commented:
@nitinmukesh `enable_sequential_cpu_offload()` should work with any diffusers model out of the box. Are you facing any particular error? Apologies about iterating slow on this PR as there are many integrations happening in parallel, but will try to wrap it up soon to fix any sequential offloading problem you might be facing

bubbliiiing commented:
> enable_sequential_cpu_offload

It seems that qwen2vl does not support enable_sequential_cpu_offload and will prompt the following error message:
![image](https://github.com/user-attachments/assets/5b98d943-f7ab-4566-8324-22bf67d9473f)

nitinmukesh commented:
Exactly this is the error I'm getting.
Currently with int4 and enable_model_cpu_offload it is working fine for 10 GB VRAM.
With sequential it should work for as low as 6 GB VRAM. However if it is not supported for qwen2vl we can simply ignore adding enable_sequential_cpu_offload.

a-r-r-o-w commented:
cc @SunMarc here for sequential cpu offload related issues in Qwen2VL

nitinmukesh commented:
> @nitinmukesh `enable_sequential_cpu_offload()` should work with any diffusers model out of the box. Are you facing any particular error? Apologies about iterating slow on this PR as there are many integrations happening in parallel, but will try to wrap it up soon to fix any sequential offloading problem you might be facing

No need to aplogize, you guys are doing very good. I'm not in hurry just trying all the features that are being added and posting my feedback. I want to see this library grow as it's very easy to use.

nitinmukesh commented:
@a-r-r-o-w 

Here is the code to reproduce the error

[pipeline.enable_model_cpu_offload() works fine]

```
import torch
from diffusers import BitsAndBytesConfig as DiffusersBitsAndBytesConfig, EasyAnimateTransformer3DModel, EasyAnimatePipeline
from diffusers.utils import export_to_video
from transformers import AutoModel
dtype = torch.bfloat16
quant_config = DiffusersBitsAndBytesConfig(load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_compute_dtype=dtype)
text_encoder_4bit = AutoModel.from_pretrained(
    "alibaba-pai/EasyAnimateV5.1-12b-zh",
    subfolder="text_encoder",
    quantization_config=quant_config,
    torch_dtype=torch.bfloat16,
)
transformer_4bit = EasyAnimateTransformer3DModel.from_pretrained(
    "alibaba-pai/EasyAnimateV5.1-12b-zh",
    subfolder="transformer",
    quantization_config=quant_config,
    torch_dtype=dtype,
)
pipeline = EasyAnimatePipeline.from_pretrained(
    "alibaba-pai/EasyAnimateV5.1-12b-zh",
    text_encoder=text_encoder_4bit,
    transformer=transformer_4bit,
    torch_dtype=dtype,
)
pipeline.enable_sequential_cpu_offload()
pipeline.vae.enable_slicing()
pipeline.vae.enable_tiling()
prompt = "A cat walks on the grass, realistic style."
negative_prompt = "bad detailed"
video = pipeline(
    prompt=prompt, 
    negative_prompt=negative_prompt, 
    num_frames=81, 
    num_inference_steps=40,
    width=512,
    height=320
).frames[0]
export_to_video(video, "cat2.mp4", fps=8)

```

and log

Log 1: https://github.com/huggingface/diffusers/pull/10626#issuecomment-2653335379

Log 2:
```
(venv) C:\aiOWN\diffuser_webui>python easyanimate_bnb.py
`low_cpu_mem_usage` was None, now default to True since model is quantized.
Downloading shards: 100%|███████████████████████████████████████████████████| 5/5 [00:00<?, ?it/s]
`Qwen2VLRotaryEmbedding` can now be fully parameterized by passing the model config through the `config` argument. All other arguments will be removed in v4.46
Loading checkpoint shards: 100%|████████████████████████████████████| 5/5 [00:34<00:00,  6.83s/it]
The config attributes {'add_ref_latent_in_control_model': True, 'clip_channels': None, 'enable_clip_in_inpaint': False, 'ref_channels': None, 'swa_layers': None} were passed to EasyAnimateTransformer3DModel, but are not expected and will be ignored. Please verify your config.json configuration file.
Expected types for text_encoder: ['Qwen2VLForConditionalGeneration', 'BertModel'], got Qwen2VLModel.
Loading pipeline components...:   0%|                                       | 0/5 [00:00<?, ?it/s]The config attributes {'force_upcast': True, 'mid_block_use_attention': False, 'sample_size': 256, 'slice_mag_vae': False, 'slice_compression_vae': False, 'cache_compression_vae': False, 'cache_mag_vae': True, 'use_tiling': False, 'norm_type': None, 'use_tiling_encoder': False, 'use_tiling_decoder': False, 'mid_block_attention_type': 'spatial'} were passed to AutoencoderKLMagvit, but are not expected and will be ignored. Please verify your config.json configuration file.
Loading pipeline components...: 100%|███████████████████████████████| 5/5 [00:01<00:00,  4.86it/s]
Traceback (most recent call last):
  File "C:\aiOWN\diffuser_webui\easyanimate_bnb.py", line 30, in <module>
    video = pipeline(
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\torch\utils\_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\diffusers\pipelines\easyanimate\pipeline_easyanimate.py", line 803, in __call__
    ) = self.encode_prompt(
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\diffusers\pipelines\easyanimate\pipeline_easyanimate.py", line 383, in encode_prompt
    prompt_embeds = text_encoder(
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\torch\nn\modules\module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\torch\nn\modules\module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\accelerate\hooks.py", line 176, in new_forward
    output = module._old_forward(*args, **kwargs)
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\transformers\models\qwen2_vl\modeling_qwen2_vl.py", line 1082, in forward
    inputs_embeds = self.embed_tokens(input_ids)
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\torch\nn\modules\module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\torch\nn\modules\module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\accelerate\hooks.py", line 171, in new_forward
    args, kwargs = module._hf_hook.pre_forward(module, *args, **kwargs)
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\accelerate\hooks.py", line 370, in pre_forward
    return send_to_device(args, self.execution_device), send_to_device(
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\accelerate\utils\operations.py", line 174, in send_to_device
    return honor_type(
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\accelerate\utils\operations.py", line 81, in honor_type
    return type(obj)(generator)
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\accelerate\utils\operations.py", line 175, in <genexpr>
    tensor, (send_to_device(t, device, non_blocking=non_blocking, skip_keys=skip_keys) for t in tensor)
  File "C:\aiOWN\diffuser_webui\venv\lib\site-packages\accelerate\utils\operations.py", line 155, in send_to_device
    return tensor.to(device, non_blocking=non_blocking)
NotImplementedError: Cannot copy out of meta tensor; no data!

```

bubbliiiing commented:
@a-r-r-o-w 
Hi, I have re-uploaded the prediction code and pretrained weights because the norm_added_q in diffusers' attention module originally did not support nn.LayerNorm, making it inconvenient for us to directly overwrite the weights. Could you please test the following prediction code for me?
```python
import torch
import numpy as np
from diffusers import EasyAnimatePipeline
from diffusers.utils import export_to_video

pipe = EasyAnimatePipeline.from_pretrained("alibaba-pai/EasyAnimateV5.1-7b-zh-diffusers", torch_dtype=torch.float16)

pipe.enable_model_cpu_offload()
pipe.vae.enable_tiling()
pipe.vae.enable_slicing()

prompt = (
    "A panda, dressed in a small, red jacket and a tiny hat, sits on a wooden stool in a serene bamboo forest. "
    "The panda's fluffy paws strum a miniature acoustic guitar, producing soft, melodic tunes. Nearby, a few other "
    "pandas gather, watching curiously and some clapping in rhythm. Sunlight filters through the tall bamboo, "
    "casting a gentle glow on the scene. The panda's face is expressive, showing concentration and joy as it plays. "
    "The background includes a small, flowing stream and vibrant green foliage, enhancing the peaceful and magical "
    "atmosphere of this unique musical performance."
)
negative_prompt = "bad detailed"
height = 512
width = 512
guidance_scale = 6
num_inference_steps = 50
num_frames = 49
seed = 43
generator = torch.Generator(device="cuda").manual_seed(seed)

video = pipe(
    prompt=prompt, 
    negative_prompt=negative_prompt,
    guidance_scale=guidance_scale, 
    num_inference_steps=num_inference_steps,
    num_frames=num_frames,
    height=height,
    width=width,
    generator=generator,
).frames[0]
print(np.shape(video))
export_to_video(video, "output.mp4", fps=8)
```

a-r-r-o-w commented:
Thank you for making the processor related changes @bubbliiiing! Will check and try to wrap this up quickly

a-r-r-o-w commented:
@bubbliiiing I've verified that the `-diffusers` suffixed checkpoints works with the T2V pipeline. I've refactored the VAE to mostly diffusers-expected format. Only the pipelines remain now, specifically the following changes:
- Removing einops dependency
- Creating a specific RoPE layer similar to how we do in [HunyuanVideo](https://github.com/huggingface/diffusers/blob/b0550a66cc3c882a1b88470df7e26103208b13de/src/diffusers/models/transformers/transformer_hunyuan_video.py#L322) and other latest model integrations
- Removing calls to `Image.open`. The users will always have to pass a PIL.Image.Image, np.array or torch.Tensor themselves
- Some other minimal cleanup

I can take these up when I find some time this week, but if you or someone from your team has the time to look into this, it would be very helpful too!

<table align=center>
<td align=center>
<video src="https://github.com/user-attachments/assets/2c3f03e6-9b11-4844-a9b6-2fdf432989fc"> Your browser does not support the video tag. </video>
</td>
</table>

bubbliiiing commented:
Sure, I'll do my best to make the necessary fixes. If there are any parts I haven't covered, please let me know and I'll be happy to help with the modifications. Thank you!

bubbliiiing commented:
I have fixed some issues. Please take another look when you have time.

a-r-r-o-w reviewed (approved):
Thanks @bubbliiiingm, the latest changes are great! @yiyixuxu Could you give this a look?

a-r-r-o-w commented on src/diffusers/pipelines/easyanimate/pipeline_easyanimate.py:
```diff
@@ -0,0 +1,964 @@
+        vae: AutoencoderKLMagvit,
+        text_encoder: Union[Qwen2VLForConditionalGeneration, BertModel],
+        tokenizer: Union[Qwen2Tokenizer, BertTokenizer],
+        text_encoder_2: Optional[Union[T5EncoderModel, Qwen2VLForConditionalGeneration]],
+        tokenizer_2: Optional[Union[T5Tokenizer, Qwen2Tokenizer]],
```
I completely missed these. EasyAnimate only uses a single text encoder, right? So we should remove these and update encode_prompt methods accordingly.

bubbliiiing replied:
Yes, do I need to delete text_encoder_2?

a-r-r-o-w commented on src/diffusers/pipelines/easyanimate/pipeline_easyanimate_control.py:
```diff
@@ -0,0 +1,1190 @@
+        )
+        self.video_processor = VideoProcessor(vae_scale_factor=self.vae_spatial_compression_ratio)
+
+    def encode_prompt(
```
As this is the same for all easyanimate pipelines, let's add a `# Copied from`

bubbliiiing replied:
Thanks.

a-r-r-o-w commented on src/diffusers/pipelines/easyanimate/pipeline_easyanimate_control.py:
```diff
@@ -0,0 +1,1190 @@
+            control_image_latents = None
+
+        return control, control_image_latents
+
+    def decode_latents(self, latents):
+        latents = 1 / self.vae.config.scaling_factor * latents
+        video = self.vae.decode(latents).sample
+        return video
```
```suggestion
```

This logic has been moved to the pipeline already (for consistency) in my latest commit but forgot to remove here

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_10626). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

yiyixuxu reviewed (approved):
awesome! thank you

yiyixuxu commented:
we have some tests are still failing though,
